### ISSUE TITLE
[Question] Is A2C deterministic on training?

### ISSUE THREAD

**Issue by xicocaio** [DLR-RM/stable-baselines3#557]
The [documentation](https://stable-baselines3.readthedocs.io/en/master/modules/a2c.html#id1) states that A2C is deterministic:

> A synchronous, deterministic variant of Asynchronous Advantage Actor Critic (A3C). It uses multiple workers to avoid the use of a replay buffer.

However, the `learn` method from the `OnPolicyAlgorithm` class collects rollouts through the `forward` method that calls `ActorCriticPolicy` with `deterministic= False`, as follows.

```
class OnPolicyAlgorithm(BaseAlgorithm):
    def learn(...) -> "OnPolicyAlgorithm":
            (...)
            continue_training = self.collect_rollouts(self.env, callback, self.rollout_buffer, n_rollout_steps=self.n_steps)

    def collect_rollouts(self, env: VecEnv, callback: BaseCallback, rollout_buffer: RolloutBuffer, n_rollout_steps: int) -> bool:
                (...)
                actions, values, log_probs = self.policy.forward(obs_tensor)

class ActorCriticPolicy(BasePolicy):
    (...)
    def forward(self, obs: th.Tensor, deterministic: bool = False) -> Tuple[th.Tensor, th.Tensor, th.Tensor]:
        (...)
        actions = distribution.get_actions(deterministic=deterministic)
```

So it appears that A2C is sampling actions through a stochastic policy during the learning phase, as the [original paper](https://arxiv.org/abs/1602.01783) describes. Also, it seems that A2C is only optionally deterministic when calling the `predict` method and that A2C.

Am I missing something? Is the A2C implementation really deterministic?

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by araffin** [comment]
Hello,

> Am I missing something? Is the A2C implementation really deterministic?

There is probably a confusion about deterministic results and deterministic policy.
A2C has deterministic results but uses a stochastic policy to collect samples.

See how we test the determinism (same random seed should lead to same results even though we sample pseudo-random numbers):
https://github.com/DLR-RM/stable-baselines3/blob/1afc2f3abe09cd4c16a6a70e16b08e9e6d138ea1/tests/test_deterministic.py#L11

**Comment by xicocaio** [comment]
Hello @araffin,

I just read `test_deterministic_training_common`, but now I don't understand the `deterministic` flag in the `predict` method. Once the seed is fixed, will it always yield the same actions under given states?

Just for clarification to see if I understand, A2C is considered deterministic even if its implementation uses stochastic policies because it yields deterministic results given that seeds are unchanged. But then, given that DPG and DDPG are considered deterministic algorithms and employ a deterministic policy during learning, what would be a truly stochastic algorithm in SB3? For instance, how would a stochastic algorithm such as REINFORCE be implemented in SB3?

**Comment by araffin** [comment]
> but now I don't understand the deterministic flag in the predict method. 

It seems that you should read more about policy gradient and the underlying probability distribution that they use: https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/distributions.py (see `sample()` -> `deterministic=False` vs `mode()` -> `deterministic=True`)

> given that DPG and DDPG are considered deterministic algorithms and employ a deterministic policy during learning,

DDPG rely on the deterministic policy gradient but still use a stochastic policy for exploration (by adding external noise).

>how would a stochastic algorithm such as REINFORCE be implemented in SB3?

The same way as A2C: https://github.com/osigaud/stable-baselines3/blob/master/stable_baselines3/reinforce/reinforce.py (this implementation is for an RL course)

**Comment by xicocaio** [comment]
> from a trained model, you can have a deterministic or stochastic policy.

Sure, this is the reason I am having trouble understanding, given that ( `deterministic=False` -> `sample()` and  `deterministic=True` -> `mode()`), why results are expected to be the same in `test_deterministic_training_common` when `model.predict(obs, deterministic=False)` is called?

**Comment by araffin** [comment]
> why results are expected to be the same in test_deterministic_training_common when model.predict(obs, deterministic=False) is called?

results of two different runs but starting with the same pseudo-random seed, so the pseudo-random sequence will be the same.

**Comment by xicocaio** [comment]
> results of two different runs but starting with the same pseudo-random seed, so the pseudo-random sequence will be the same.

I understand that a seed influences initialization and other aspects that lead to the final trained model. But to make sure that I comprehend what you are pointing out to me.

Lets say that after training a stochastic RL algorithm, the categorical probabilities of taking actions at a given state S are P(A|S) = 85%, P(B|S) = 15%, P(C|S) = 5%. Then, if I want this model to behave deterministically after training, it would select A every time it got to that given state S. But, if I want a stochastic behavior, it would selected action A 85% of the time when reaching state S.

Then the confusing part for me is that during the prediction phase, even though `deterministic=False`, actions are always the same for models trained with the same seed.

In this sense, if I want a truly stochastic behavior during prediction, shouldn't `torch.manual_seed(random_seed)` be called before `distribution.sample()` in order to set different seed and guarantee stochastic sampling?

**Comment by Miffyli** [comment]
Just to pitch in here: I think the whole "deterministic" part  in the docs regarding A2C is confusing. I do not see how A2C is "deterministic" in any way (sure, the results match if you fix seed, but that applies to all algorithms once you fix the PRNG seed for everything).

> In this sense, if I want a truly stochastic behavior during prediction, shouldn't torch.manual_seed(random_seed) be called before distribution.sample() in order to set different seed and guarantee stochastic sampling?

You should not need to set the seed manually (and should not set it to anything fixed! Otherwise you will be stuck with deterministic behaviour). Simply doing `predict` with `deterministic=False` should be enough.

**Comment by xicocaio** [comment]
@Miffyli Exactly, agree 100%.

Moreover, the way I see it, currently, the `deterministic` flag has no real use and leads to the same `predict()` behavior.

**Comment by araffin** [comment]
> Then the confusing part for me is that during the prediction phase, even though deterministic=False, actions are always the same for models trained with the same seed.

I think there is a confusion about what the test checks:

https://github.com/DLR-RM/stable-baselines3/blob/1afc2f3abe09cd4c16a6a70e16b08e9e6d138ea1/tests/test_deterministic.py#L34

it checks that the sequence of actions (and therefore the sequence of rewards) is the same between two runs. Successive calls to `predict(obs, deterministic=False)` with the same observation lead to different actions, sampled according to the learned proba distribution.

>Moreover, the way I see it, it seems that the deterministic flag currently has no real use and leads to the same predict() behavior.

by default, `deterministic=False`, so predict with and without that flag is the same. However, it is good practice to explicitly specify it as the difference is quite important (for instance for continuous control).

**Comment by xicocaio** [comment]
I understand from the test that two models are trained with the same parameters on a given env. Then, this env is reset, and the models are subject to this env again and yield the same sequence St At Rt St+1 At+1 Rt+1... with `deterministic=False`.

I see that the model is learning a proba distribution, but according to this test either the proba is always converging to a given action (e.g., P(A|St=s) = 100%), or it is always deterministically choosing the same action given heterogenous probabilities (e.g., P(A|St=s) = 50%, P(B|St=s) = 30%, P(ACSt=s) = 20%).

In this sense, it seems only deterministic behaviors are occurring. Thus, is there a way to achieve a stochastic behavior?

I would expect that in a stochastic scenario, `predict(deterministic=False)` would lead to different sequences St A' R' S'' A'' R''...

Hence, these two sentences seem to be contradictory.

> it checks that the sequence of actions (and therefore the sequence of rewards) is the same between two runs.

> Successive calls to predict(obs, deterministic=False) with the same observation lead to different actions, sampled according to the learned proba distribution.

The test confirms that the model yields the same sequence of actions and rewards for two models trained with the same parameters. So, how can successive calls to `predict()` with the same observation lead to different actions?

**Comment by araffin** [comment]
> So, how can successive calls to predict() with the same observation lead to different actions?

because we are sampling from the same learned distribution starting with a pseudo-random generator we initialized in the same random state? (as we used the same initial random seed)

To elaborate, after the training, we have two models that have the same parameters and we have two pseudo-random generators that are in the same state (hidden in numpy/pytorch).
Successive calls to `predict()` will sample actions from the learned distribution.
As the two models are the same, the two learned distributions are the same.
And because the two pseudo-random generators are the same, the sequence of sampled actions will be the same.

**Comment by araffin** [comment]
> Yes, the trained A2C model is deterministic.

in a way yes, as all computer programs not involving non-deterministic computation such as threading without locks.

>The deterministic argument changes how the trained model selects actions as such 

yes

>However, in practice, the model will work either way deterministically

I'm not sure to follow, best is to try by yourself:

```python
from stable_baselines3 import A2C

# Set the seed, you can also pass `seed=None` to have different results at every run
model = A2C("MlpPolicy", "Pendulum-v0", seed=1)
obs = model.get_env().reset()

for _ in range(10):
    print(model.predict(obs, deterministic=False)[0])

**Comment by xicocaio** [comment]
Running your code with `deterministic=True` but with `range(3)` for simplification, we have:

```
deterministic=True              deterministic=Falsae
[[0.00128911]]                     [[-0.2949889]]
[[0.00128911]]                     [[-0.6326946]]  
[[0.00128911]]                     [[-0.5796663]]
```

Sure, they are different, but my issue is that the random state does not constantly reset and, thus, we get the same sequence as before if `deterministic=False`.

```
[[-0.2949889]]
[[-0.6326946]]
[[-0.5796663]]
```

I understand that for reproducibility requirements of most research works, libraries such as PyTorch do not reset their random state at each application run. So we get the same sequence every time we rerun the application.

However, I was expecting a behavior similar to a typical python application that resets the random state at each run (`python app.py`), as such:

```
import numpy as np

for _ in range(3):
    print(np.random.randint(1000))

1st run              2nd run
56                     446
316                    3
431                    588
```

In this sense, to achieve the desired behavior, I had to reset the random state manually.

```
import torch
from numpy.random import RandomState

model = A2C("MlpPolicy", "Pendulum-v0", seed=1)
obs = model.get_env().reset()

prng = RandomState()

for _ in range(3):
    torch.manual_seed(prng.randint(1000))
    print(model.predict(obs, deterministic=False)[0])

first run                              second run
[[-0.4679766]]                      [[-1.0295607]]
[[-0.20940204]]                     [[1.0567384]]
[[-1.315447]]                       [[1.4763789]]

```

In conclusion, I expected that by default, the different application runs with `deterministic=False` would generate a non-deterministic sequence. But, instead, it is generating a deterministic series of non-deterministic actions. And that's why I agree with Miffyli's statement that,

> the whole "deterministic" part in the docs regarding A2C is confusing.

and could be improved.

**Comment by araffin** [comment]
as written in my comment, pass `seed=None` (the default) to have the behavior you expect ;)

**Comment by xicocaio** [comment]
Still, for my purpose, I need to set a specific seed for training and then reach non-deterministic series at each run. Thus, passing `seed=None` is not an option. So, I will keep manually resetting my random state as shown previously.

Anyway, thank you for your time.
